### PR TITLE
[upmeter] Tweak agent HTTP timeout

### PR DESCRIPTION
**Pull request by shvgn**
## Description

Set more sane timeouts to upmeter-agent sender.

## Why do we need it, and what problem does it solve?

Reduces dummy logs about reached timeout and request cancellations.

## What is the expected result?

Reduced or diminished logging or cancelled requests in upmeter-agent

## Checklist
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
```changes
section: upmeter
type: fix
summary: Tweak HTTP sending timeouts to reduce cancellation logs
impact_level: low
```

### CODE REVIEW

File: modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go [hunk 1]
```diff
@@ -142,15 +144,17 @@ func createHttpTransport(caPath string) (*http.Transport, error) {
...
-		TLSClientConfig: &tls.Config{
-			RootCAs: caCertPool,
-		},
-	}
+	tr.TLSClientConfig = &tls.Config{RootCAs: caCertPool}
 
 	return tr, nil
 }
 
+func newTransport(timeout time.Duration) *http.Transport {
```
Comment (nabokihms) [inline modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go:152]: It is unclear that the timeout argument is actually a ResponseHeaderTimeout.

File: modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go [hunk 2]
```diff
@@ -89,11 +90,14 @@ func (c *Client) Send(reqBody []byte) error {
...
-func NewHttpClient(config *ClientConfig, timeout time.Duration) *http.Client {
-	client, err := createSecureHttpClient(config.TLS, config.CAPath, timeout)
+func NewHttpClient(config *ClientConfig) *http.Client {
+	client, err := createSecureHttpClient(config.TLS, config.CAPath, config.Timeout)
 	if err != nil {
 		log.Errorf("falling back to default HTTP client: %v", err)
-		return &http.Client{Timeout: timeout}
+		return &http.Client{
+			Timeout:   config.Timeout,
+			Transport: newTransport(config.Timeout / 2),
```
Comment (nabokihms) [inline modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go:99]: 1. Why do we set timeout / 2 to read headers? For example, if there are only headers in the response, it's ok to spend the whole time reading headers.

2. Shouldn't other timeouts be set, e.g., connection timeout, handshake timeout?

3. There is the same division (timeout / 2) in two places. Maybe it is better to put this logic into the `newTransport` function? It makes more sense to me cos we will accept some generic timeout and then describe the logic of why we use `timeout / 2` for headers inside the function.

File: modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go [hunk 3]
```diff
@@ -94,10 +94,15 @@ func parseAgentArgs(cmd *kingpin.CmdClause, config *agent.Config) {
 		Default("false").
 		BoolVar(&config.ClientConfig.TLS)
 
-	cmd.Flag("period", "The period of episodes sending to server, and at the same the client timeout.").
-		Envar("UPMETER_PERIOD").
+	cmd.Flag("send-interval", "The period of episodes sending to server, and at the same the client timeout.").
```
Comment (nabokihms) [inline modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go:97]:
```suggestion
	cmd.Flag("send-interval", "The period of sending episodes to the upmeter server.").
```

File: modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go [hunk 4]
```diff
@@ -94,10 +94,15 @@ func parseAgentArgs(cmd *kingpin.CmdClause, config *agent.Config) {
...
+		Envar("UPMETER_SEND_INTERVAL").
 		Default("1s").
-		DurationVar(&config.Period)
+		DurationVar(&config.Interval)
+
+	cmd.Flag("send-timeout", "The period of episodes sending to server, and at the same the client timeout.").
```
Comment (nabokihms) [inline modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go:102]:
```suggestion
	cmd.Flag("send-timeout", "The timeout of sending episodes to the upmeter server.").
```

File: modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go (hunk 1 above)
Comment (shvgn) [reply modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go:158]: Let it be just "a transport timeout"! it is reused in multiple places.

File: modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go (hunk 2 above)
Comment (shvgn) [reply modules/500-upmeter/images/upmeter/pkg/agent/sender/client.go]: 1. Agreed. I removed the division.
2. I agree. That's why there was a copy of http.DefaultClient, where all the timeouts are set. And I changed only one that matters. It is done differently now, anyway.
3. I'd rather keep it explicit in two places than changing passed value implicitly under the hood. You pass a value, but it is halved! Anyway, I removed the division due to the first point.

File: modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go (hunk 3 above)
Comment (shvgn) [reply modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go]: The text has changed

File: modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go (hunk 4 above)
Comment (shvgn) [reply modules/500-upmeter/images/upmeter/cmd/upmeter/parse_args.go]: The text has changed

**Review by nabokihms — APPROVED** [review]

**Pull request merged**